ccl/sqlproxyccl/balancer: TestRebalancer_processQueue failed

ccl/sqlproxyccl/balancer.TestRebalancer_processQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4913679&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4913679&tab=artifacts#/) on master @ [dcb06e12a77c0a7f7fe9d2609c54327345987dcd](https://github.com/cockroachdb/cockroach/commits/dcb06e12a77c0a7f7fe9d2609c54327345987dcd):


```
=== RUN   TestRebalancer_processQueue
--- FAIL: TestRebalancer_processQueue (0.77s)
=== RUN   TestRebalancer_processQueue/retries_up_to_maxTransferAttempts
    balancer_test.go:119: 
        	Error Trace:	balancer_test.go:119
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestRebalancer_processQueue/retries_up_to_maxTransferAttempts
    --- FAIL: TestRebalancer_processQueue/retries_up_to_maxTransferAttempts (0.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRebalancer_processQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/balancer.TestRebalancer_processQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4916565&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4916565&tab=artifacts#/) on master @ [6c91da48b8da78c13e4cb8675a1a98c5ec75ea06](https://github.com/cockroachdb/cockroach/commits/6c91da48b8da78c13e4cb8675a1a98c5ec75ea06):


```
=== RUN   TestRebalancer_processQueue
--- FAIL: TestRebalancer_processQueue (0.76s)
=== RUN   TestRebalancer_processQueue/conn_was_transferred
    balancer_test.go:172: 
        	Error Trace:	balancer_test.go:172
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestRebalancer_processQueue/conn_was_transferred
    --- FAIL: TestRebalancer_processQueue/conn_was_transferred (0.00s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRebalancer_processQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/sqlproxyccl/balancer/balancer_test.go
+++ b/pkg/ccl/sqlproxyccl/balancer/balancer_test.go
@@ -94,6 +94,18 @@ func TestRebalancer_processQueue(t *testing.T) {
 		dst: "foo",
 	}
 
+	// assertNoRunningRequests asserts that the rebalance loop isn't processing
+	// any request.
+	assertNoRunningRequests := func(t *testing.T) {
+		testutils.SucceedsSoon(t, func() error {
+			runningReq := b.metrics.rebalanceReqRunning.Value()
+			if runningReq != 0 {
+				return errors.Newf("expected no running requests, but got %d", runningReq)
+			}
+			return nil
+		})
+	}
+
 	t.Run("retries_up_to_maxTransferAttempts", func(t *testing.T) {
 		count := 0
 		req := &rebalanceRequest{
@@ -114,9 +126,10 @@ func TestRebalancer_processQueue(t *testing.T) {
 		b.queue.enqueue(syncReq)
 		<-syncCh
 
+		assertNoRunningRequests(t)
+
 		// Ensure that we only retried up to 3 times.
 		require.Equal(t, 3, count)
-		require.Equal(t, int64(0), b.metrics.rebalanceReqRunning.Value())
 	})
 
 	t.Run("conn_was_transferred_by_other", func(t *testing.T) {
@@ -141,9 +154,10 @@ func TestRebalancer_processQueue(t *testing.T) {
 		b.queue.enqueue(syncReq)
 		<-syncCh
 
+		assertNoRunningRequests(t)
+
 		// We should only retry once.
 		require.Equal(t, 1, count)
-		require.Equal(t, int64(0), b.metrics.rebalanceReqRunning.Value())
 	})
 
 	t.Run("conn_was_transferred", func(t *testing.T) {
@@ -167,9 +181,10 @@ func TestRebalancer_processQueue(t *testing.T) {
 		b.queue.enqueue(syncReq)
 		<-syncCh
 
+		assertNoRunningRequests(t)
+
 		// We should only retry once.
 		require.Equal(t, 1, count)
-		require.Equal(t, int64(0), b.metrics.rebalanceReqRunning.Value())
 	})
 
 	t.Run("conn_was_closed", func(t *testing.T) {
@@ -192,9 +207,10 @@ func TestRebalancer_processQueue(t *testing.T) {
 		b.queue.enqueue(syncReq)
 		<-syncCh
 
+		assertNoRunningRequests(t)
+
 		// We should only retry once.
 		require.Equal(t, 1, count)
-		require.Equal(t, int64(0), b.metrics.rebalanceReqRunning.Value())
 	})
 
 	t.Run("limit_concurrent_rebalances", func(t *testing.T) {

